feat: Authenticate protected media requests on private sites

dcalhoun opened the pull request:
## Description
<!-- Describe the changes, why they are needed, and how they address the issue -->
Leverage cookie configuration provided by GutenbergKit's https://github.com/wordpress-mobile/GutenbergKit/pull/144 to authenticate requests for protected media on private sites. 

Fix CMM-452.

## Testing instructions
<!-- Consider listing specific testing steps for helping reviewers evaluate the efficacy of the changes -->
> [!Important]
> This requires applying https://github.com/Automattic/jetpack/pull/44328 to your Atomic site, as it is not yet deployed to production. 

### Verify protected media loads

1. Enable the experimental block editor.
1. Open a post containing media on a private WPCOM Simple site. 
1. Verify the media loads as expected. 
1. Repeat with a private WPCOM Atomic site. 

<!--

Consider testing the following *yourself* before requesting a review:

- Compatibility with WordPress.com sites and self-hosted Jetpack sites
- Both portrait and landscape orientations
- Light and dark modes
- Dynamic font sizes: larger, smaller and bold text
- High contrast mode
- Screen reader experience (e.g., Talkback)
- Languages with large words or with letters/accents not frequently used in English
- Right-to-left languages layout
- Multi-tasking: split view and slide over

-->

<!--

Use these emoji to improve the code review

❤️ - Praise. Compliments about great solution.
⛏ - Nitpick️. Not high importance. Usually, short one-liners.
💡 - Idea. Alternative solution or improvement.
❓ - Question. Requires some explanation or more context.
⚠️ - Warning. Should be addressed in this PR or in some follow-up PR.
🛑 - Blocker. Must be addressed before PR gets merged.

-->

dcalhoun committed:
feat: Authenticate protected media requests on private Atomic sites

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
style: Address lint warning

oguzkocer committed:
Fix WordPressCookieAuthenticator to properly use CookieJar for authentication

- Use CookieJar to retrieve cookies instead of parsing response headers
- Extract cookies from jar using loadForRequest() for proper cookie management
- Add createDefaultOkHttpClient() with JavaNetCookieJar configuration
- Support both dependency injection and direct instantiation patterns
- Maintain comprehensive debug logging for cookie authentication flow

This should resolve the issue where only wordpress_test_cookie was received
by properly leveraging OkHttp's cookie management infrastructure.

dcalhoun committed:
feat: Configure cookies for private WPCOM Simple sites

Ensure protected media loads.

dcalhoun committed:
refactor: Address Detekt errors

dcalhoun committed:
fix: Address build errors

dcalhoun committed:
fix: Address build errors

dcalhoun committed:
fix: Implement rudimentary okhttp3 cookie jar

Enable retrieving cookies for GutenbergKit requests.

dcalhoun committed:
style: Fix lint warnings

dcalhoun committed:
test: Assert WordPressCookieAuthenticator expectations

dcalhoun committed:
refactor: Address lint warnings

dcalhoun committed:
refactor: Remove unnecessary comments and empty lines

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/modules/ApplicationModule.java:
```diff
@@ -154,4 +166,46 @@ public static RecordingStrategy provideVoiceToContentRecordingStrategy() {
+    @Singleton
+    public static OkHttpClient provideOkHttpClient() {
+        // Create simple CookieJar for cookie authentication
+        CookieJar cookieJar = new CookieJar() {
+            private final List<Cookie> mCookies = new ArrayList<>();
+            
+            @Override
+            public void saveFromResponse(@NonNull HttpUrl url, @NonNull List<Cookie> cookies) {
+                this.mCookies.addAll(cookies);
+            }
+            
+            @NonNull
+            @Override
+            public List<Cookie> loadForRequest(@NonNull HttpUrl url) {
+                List<Cookie> validCookies = new ArrayList<>();
+                for (Cookie cookie : mCookies) {
+                    if (cookie.matches(url)) {
+                        validCookies.add(cookie);
+                    }
+                }
+                return validCookies;
+            }
+        };
+        
+        // OkHttpClient with proper cookie handling
+        return new OkHttpClient.Builder()
+                .cookieJar(cookieJar)
+                .connectTimeout(15, TimeUnit.SECONDS)
+                .readTimeout(15, TimeUnit.SECONDS)
+                .writeTimeout(15, TimeUnit.SECONDS)
+                .followRedirects(true)
+                .followSslRedirects(true)
+                .build();
+    }
+
+    @Provides
+    @Singleton
+    public static CoroutineScope provideCoroutineScope() {
+        return CoroutineScopeKt.CoroutineScope(Dispatchers.getDefault());
+    }
```
This was introduced in https://github.com/wordpress-mobile/WordPress-Android/pull/22047/commits/92414633736a3f541181c93b7f3ca7529213dc32 to address the following build errors. 

```
Type org.wordpress.android.editor.gutenberg.WordPressCookieAuthenticator may only contain one injected constructor. Found: [@Inject org.wordpress.android.editor.gutenberg.WordPressCookieAuthenticator(okhttp3.OkHttpClient, kotlinx.coroutines.CoroutineScope), @Inject org.wordpress.android.editor.gutenberg.WordPressCookieAuthenticator()] - org.wordpress.android.editor.gutenberg.WordPressCookieAuthenticator
```

Afterwards, a simple cookie jar was created in https://github.com/wordpress-mobile/WordPress-Android/pull/22047/commits/3431eab029691ed8ad8d3c9fbae2ba36292eef9a as the cookies were missing in the new implementation. 

Some of the lingering, unused, original logic was removed in https://github.com/wordpress-mobile/WordPress-Android/pull/22047/commits/3854a2f4cb57bf39db0306f4bd186692aa737c24. 

I request close review of these specific changes. It feels odd to hoist this logic so high in the application structure. I welcome alternative approaches.

oguzkocer replied:
These don't look quite right to me. I'll revert the change and simplify the logic in a follow up PR.

dcalhoun reviewed (commented):
@oguzkocer I integrated the changes from https://github.com/wordpress-mobile/WordPress-Android/commit/def922e2960fa6257bf294b85b3b84a3ebe5fc2d. I modified the implementation with Claude's help to address build errors noted within inline comments. I'd like your thoughts on those changes specifically. 

Would you please review these proposed changes?

oguzkocer committed:
refactor: Move WordPressCookieAuthenticator to WordPress module

Move WordPressCookieAuthenticator from libs:editor to WordPress/auth to enable dependency injection.

Changes:
- Move class from libs/editor to WordPress/auth package
- Update to use @Named dependency injection for OkHttpClient and CoroutineDispatcher
- Update EditPostActivity import
- Remove unused ApplicationModule providers
- Simplify tests to only include parameter validation

oguzkocer committed:
Remove no longer necessary libs/editor/build.gradle dependency additions

oguzkocer committed:
refactor: Extract authentication logic to ViewModel with sealed classes

Moved cookie authentication from Activity to dedicated ViewModel for better
architecture and state management.

Changes:
- Extract EditPostAuthViewModel with sealed class states
- Remove callback complexity from WordPressCookieAuthenticator
- Use coroutines and postValue() for thread safety
- Consolidate cookie logic with single source of truth
- Fix Idle state handling to prevent premature UI setup
- Use functional programming patterns for cookie processing

This improves maintainability, testability, and follows modern Android
architecture patterns with proper separation of concerns.

oguzkocer committed:
refactor: Simplify cookie logic and add comprehensive unit tests

Refactor getCookiesForPrivateSites method to use cleaner when expression
that prioritizes checking if site is private first, making the logic more
readable and maintainable.

Add comprehensive unit tests for EditPostAuthViewModel covering:
- Authentication state transitions (Loading, Success, Error)
- Cookie processing for different site types (WPCom, atomic, private/public)
- Edge cases including authentication failures and missing cookies
- Cookie filtering to only include wordpress_logged_in cookies

Changes:
- Simplify conditional logic in getCookiesForPrivateSites method
- Add EditPostAuthViewModelTest with 11 comprehensive test cases
- Use mock-based approach for PrivateAtomicCookie in tests

oguzkocer committed:
feat: Add caching and logout cleanup to WordPressCookieAuthenticator

- Implement cookie caching with 6-hour expiration using SharedPreferences
- Add cache invalidation on corrupted data and expired entries
- Include logout cleanup in AppInitializer to clear cached cookies
- Update logging to use API tag and specify WordPress.com account cookies
- Fix unit tests to properly mock HTTP requests instead of making real network calls
- Add comprehensive test coverage for caching scenarios, HTTP errors, and network failures

Changes:
- Add @Singleton annotation and PreferenceUtilsWrapper dependency
- Implement getCachedCookies(), cacheCookies(), clearCachedCookies(), and clearAllCachedCookies() methods
- Add cache-first logic in authenticateForCookies() with automatic caching of successful results
- Integrate clearAllCachedCookies() call in AppInitializer.removeWpComUserRelatedData()
- Replace real OkHttpClient with mocked version in tests to prevent actual network requests
- Add 3 new test cases covering successful authentication, HTTP errors, and network failures

oguzkocer committed:
refactor: Extract request execution to separate method in WordPressCookieAuthenticator

- Extract HTTP request execution logic from authenticateForCookies() into performAuthenticationRequest() helper method
- Add ValidatedAuthParams data class to represent validated parameters passed to helper method
- Improve separation of concerns: main method handles validation and caching orchestration, helper method handles HTTP execution
- Maintain existing functionality with clearer code organization

Changes:
- Add ValidatedAuthParams private data class for validated parameter passing
- Create performAuthenticationRequest() method containing HTTP request logic
- Update authenticateForCookies() to validate parameters then delegate to helper method
- All tests continue to pass, confirming refactoring preserves behavior

oguzkocer committed:
feat: Extend cookie cache expiration from 6 hours to 6 months

WordPress.com authentication cookies are now cached for 6 months instead of 6 hours to provide better user experience by reducing authentication requests.

oguzkocer committed:
Fix flaky test that caused uncaught exceptions in CI

The test `fetchWpComCookies starts with Loading state` was causing
NoWhenBranchMatchedException to leak into other tests due to missing
mock setup for the authenticator.

The test was also fundamentally flawed because it tried to check the
Loading state immediately after calling fetchWpComCookies(), but
postValue() is asynchronous and the mocked authenticator would complete
immediately, potentially skipping the Loading state.

Changes:
- Renamed test to better reflect what it actually tests
- Added proper mock setup to prevent exception leakage
- Changed approach to observe all state changes using observeForever
- Now validates the complete flow: Loading → Success states
- Uses advanceUntilIdle() for proper async test handling

dcalhoun committed:
Refactor `EditPostActivity` Authentication and Enhance Cookie Authenticator (#22073)

dcalhoun committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into feat/authenticate-editor-web-view-requests-for-private-sites

dcalhoun commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt:
```diff
@@ -431,6 +431,7 @@ class EditPostActivity : BaseAppCompatActivity(), EditorFragmentActivity, Editor
     @Inject lateinit var editorJetpackSocialViewModel: EditorJetpackSocialViewModel
     private lateinit var editPostNavigationViewModel: EditPostNavigationViewModel
     private lateinit var editPostSettingsViewModel: EditPostSettingsViewModel
+    @Inject lateinit var editPostAuthViewModel: EditPostAuthViewModel
```
@oguzkocer I resolved a merge conflict with the recent`@Inject`s removal from `editPostNavigationViewModel` in https://github.com/wordpress-mobile/WordPress-Android/pull/22068. I believe retaining the `@Inject` for `editPostAuthViewModel` is the correct, but I request your input on that.

oguzkocer replied:
For now, this should work fine. However, I think I'll switch it to injecting using the `ViewModelProvider` [like so](https://github.com/wordpress-mobile/WordPress-Android/blob/fff89bcedfcf2624ba867897e1cadabb6408f758/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.kt#L642-L645) in a subsequent PR. (feel free to do that in this PR if you're willing to test it again)

My understanding is that if we use `@Inject` in both the activity and another fragment, we'll get different instances of the view model. If we use the `ViewModelProvider` in both the activity, and in the fragment for example [like this](https://github.com/wordpress-mobile/WordPress-Android/blob/fff89bcedfcf2624ba867897e1cadabb6408f758/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L200-L201), the instance will be shared. This is important because it's how the communication between the VM/Activity/Fragment is established through observers.

This is one part of my plan to simplify `EditPostActivity`. First switching the logic to the VM and keep the UI logic in both activity & fragment and later maybe get rid of some of that logic. Currently we have coupling between every component, so it's impossible to refactor things in isolation. Check out #22067 & #22068 if you are interested to learn more.

dcalhoun reviewed (commented):
As noted in https://github.com/wordpress-mobile/WordPress-Android/pull/22073#pullrequestreview-3076215308, I successfully completed the testing plan for that PR. I also retested the primary functionality after merging the latest `trunk` into this branch. 

@oguzkocer will you please review this when you have time? 🙇🏻

oguzkocer reviewed (approved):
I didn't test this again, but there are no changes from my refactor PR and I've tested that extensively.
